### ISSUE TITLE
Add support for @apply custom-property-sets across the board

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#626]
So far addressed in `selector-no-vendor-prefix` by https://github.com/stylelint/stylelint/pull/625 (original issue: #624). I also checked the other selector rules that use `styleSearch()` and they are OK.

However, it looks like any rules using `postcss-selector-parser`—e.g. the `selector-*-pattern` and `selector-no-*` rules—trip over when using custom-property-sets i.e. `Error: Expected pseudo-class or pseudo-element`. Shall we address this, if so is it best to do so at our end or upstream?

Flagging this an enhancement, rather than a bug, as supporting custom-property-sets is a feature that can be currently worked around by running `postcss-apply` in the pipeline before the linter.

**Comment by davidtheclark** [comment]
cf. https://github.com/postcss/postcss-selector-parser/issues/27

**Comment by jeddy3** [comment]
So am I right in the thinking that in our case the solution is to, for each of the rules that use `postcss-selector-parser`, ignore all selectors within the `:root` pseudo-class?

I had a bit of trouble grokking the other issue and so want to double check this with you.

**Comment by davidtheclark** [comment]
Well, if something like `--my-property: pink;` is within a `:root {}` rule, then it's not a selector at all: it's a property. I *think* that the weirdness that's happening here is because people are *not* putting custom-property-sets within `:root {}` rules. The easy solution should be for them to do so. But is there some impending standard about this, some spec to look at?

**Comment by jeddy3** [comment]
> The easy solution should be for them to do so. But is there some impending standard about this, some spec to look at?

I've not found much, other than in "EXAMPLE 2" here: http://tabatkins.github.io/specs/css-apply-rule/#defining - it shows a custom-property-set being defined outside of `:root`.

I'm wondering whether it's way too early to think about supporting this syntax and instead just recommending, should this issue pop up again, running the likes of `postcss-apply` before `stylelint`?

What do you think?

**Comment by davidtheclark** [comment]
Example 2 shows them being defined outside of `:root` but still as
properties within a rule, not outside of any rule, right?

Unless I'm mistaken then, the proposed syntax is fully supported already. I
do not think we have to worry about this for now.

**Comment by jeddy3** [comment]
> Example 2 shows them being defined outside of `:root` but still as
properties within a rule, not outside of any rule, right?

Yep, you're right.

> Unless I'm mistaken then, the proposed syntax is fully supported already. 

That's what I'm trying to get at... :) I'm not sure it is already supported. All the rules that make use of `postcss-selector-parser`—e.g. the `selector-*-pattern` and `selector-no-*` rules—seem to trip up on even the `:root` syntax, e.g.

```css
:root {
  --my-custom-property-set: {
    background-color: red;
    color: blue;
  }
}
```

Throwing: `Error: Expected pseudo-class or pseudo-element`.

It looks like the custom-property-sets are being mistaken as selectors—even if they are within `:root` or another rule—and parsed as such, which leads `postcss-selector-parser` to throw the above error.

Perhaps there is a way of returning early in these rules if a custom-property-set is mistaken for a selector?

**Comment by davidtheclark** [comment]
Oh! Yes, we could look into that.
